# Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [1]:
import pandas as pd
import numpy as np

In [34]:
columns = [
    'surgery',
    'age',
    'hospital number',
    'rectal temperature',
    'pulse',
    'respiratory rate',
    'temperature of extremities',
    'peripheral pulse',
    'mucous membranes',
    'capillary refill time',
    'pain',
    'peristalsis',
    'abdominal distension',
    'nasogastric tube',
    'nasogastric reflux',
    'nasogastric reflux PH',
    'rectal examination - feces',
    'abdomen',
    'packed cell volume',
    'total protein',
    'abdominocentesis appearance',
    'abdomcentesis total protein',
    'outcome',
    'surgical lesion',
    'type of lesion(25)',
    'type of lesion(26)',
    'type of lesion(27)',
    'cp_data',
    ]    
horse_data = pd.read_csv('horse_data.csv', names=columns, na_values='?') 
# анализ инфо и уникальных значений по столбцам показал, что пропуски в данных заполнены "?", поэтому заменяем "?" на NaN

horse_data.head()

surgery  age  hospital number  rectal temperature  pulse  respiratory rate  \
0      2.0    1           530101                38.5   66.0              28.0   
1      1.0    1           534817                39.2   88.0              20.0   
2      2.0    1           530334                38.3   40.0              24.0   
3      1.0    9          5290409                39.1  164.0              84.0   
4      2.0    1           530255                37.3  104.0              35.0   

   temperature of extremities  peripheral pulse  mucous membranes  \
0                         3.0               3.0               NaN   
1                         NaN               NaN               4.0   
2                         1.0               1.0               3.0   
3                         4.0               1.0               6.0   
4                         NaN               NaN               6.0   

   capillary refill time  ...  packed cell volume  total protein  \
0                    2.0  ...                45.0            8.4   
1                    1.0  ...                50.0           85.0   
2                    1.0  ...                33.0            6.7   
3                    2.0  ...                48.0            7.2   
4                    2.0  ...                74.0            7.4   

   abdominocentesis appearance  abdomcentesis total protein  outcome  \
0                          NaN                          NaN      2.0   
1                          2.0                          2.0      3.0   
2                          NaN                          NaN      1.0   
3                          3.0                          5.3      2.0   
4                          NaN                          NaN      2.0   

   surgical lesion  type of lesion(25)  type of lesion(26)  \
0                2               11300                   0   
1                2                2208                   0   
2                2                   0                   0   
3                1                2208                   0   
4                2                4300                   0   

   type of lesion(27)  cp_data  
0                   0        2  
1                   0        2  
2                   0        1  
3                   0        1  
4                   0        2  

[5 rows x 28 columns]

In [3]:
horse_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 28 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   surgery                      299 non-null    float64
 1   age                          300 non-null    int64  
 2   hospital number              300 non-null    int64  
 3   rectal temperature           240 non-null    float64
 4   pulse                        276 non-null    float64
 5   respiratory rate             242 non-null    float64
 6   temperature of extremities   244 non-null    float64
 7   peripheral pulse             231 non-null    float64
 8   mucous membranes             253 non-null    float64
 9   capillary refill time        268 non-null    float64
 10  pain                         245 non-null    float64
 11  peristalsis                  256 non-null    float64
 12  abdominal distension         244 non-null    float64
 13  nasogastric tube    

In [4]:
len(horse_data['hospital number'].unique())

284

In [13]:
young = horse_data.query('age == 9').groupby('surgery').count()
young.head()

age  hospital number  rectal temperature  pulse  respiratory rate  \
surgery                                                                      
1.0       18               18                  15     16                15   
2.0        6                6                   5      6                 5   

         temperature of extremities  peripheral pulse  mucous membranes  \
surgery                                                                   
1.0                              13                12                13   
2.0                               3                 2                 3   

         capillary refill time  pain  ...  packed cell volume  total protein  \
surgery                               ...                                      
1.0                         14    14  ...                  17             16   
2.0                          4     3  ...                   6              6   

         abdominocentesis appearance  abdomcentesis total protein  outcome  \
surgery                                                                      
1.0                               11                            6       18   
2.0                                2                            2        6   

         surgical lesion  type of lesion(25)  type of lesion(26)  \
surgery                                                            
1.0                   18                  18                  18   
2.0                    6                   6                   6   

         type of lesion(27)  cp_data  
surgery                               
1.0                      18       18  
2.0                       6        6  

[2 rows x 27 columns]

In [36]:
abd = horse_data.groupby('abdominal distension').count()
abd.head()

surgery  age  hospital number  rectal temperature  \
abdominal distension                                                      
1.0                        75   76               76                  69   
2.0                        65   65               65                  56   
3.0                        65   65               65                  52   
4.0                        38   38               38                  23   

                      pulse  respiratory rate  temperature of extremities  \
abdominal distension                                                        
1.0                      75                62                          70   
2.0                      59                54                          57   
3.0                      60                57                          62   
4.0                      35                25                          36   

                      peripheral pulse  mucous membranes  \
abdominal distension                                       
1.0                                 68                73   
2.0                                 56                61   
3.0                                 56                62   
4.0                                 33                34   

                      capillary refill time  ...  packed cell volume  \
abdominal distension                         ...                       
1.0                                      75  ...                  74   
2.0                                      64  ...                  63   
3.0                                      61  ...                  61   
4.0                                      36  ...                  31   

                      total protein  abdominocentesis appearance  \
abdominal distension                                               
1.0                              75                           35   
2.0                              61                           36   
3.0                              62                           36   
4.0                              28                           14   

                      abdomcentesis total protein  outcome  surgical lesion  \
abdominal distension                                                          
1.0                                            28       75               76   
2.0                                            26       65               65   
3.0                                            24       65               65   
4.0                                             8       38               38   

                      type of lesion(25)  type of lesion(26)  \
abdominal distension                                           
1.0                                   76                  76   
2.0                                   65                  65   
3.0                                   65                  65   
4.0                                   38                  38   

                      type of lesion(27)  cp_data  
abdominal distension                               
1.0                                   76       76  
2.0                                   65       65  
3.0                                   65       65  
4.0                                   38       38  

[4 rows x 27 columns]

In [37]:
outcome = horse_data.groupby(['outcome', 'surgery']).count()
outcome.head(10)

age  hospital number  rectal temperature  pulse  \
outcome surgery                                                    
1.0     1.0       95               95                  80     88   
        2.0       83               83                  72     78   
2.0     1.0       58               58                  39     49   
        2.0       19               19                  14     17   
3.0     1.0       27               27                  21     27   
        2.0       17               17                  13     16   

                 respiratory rate  temperature of extremities  \
outcome surgery                                                 
1.0     1.0                    76                          75   
        2.0                    71                          71   
2.0     1.0                    45                          47   
        2.0                    13                          17   
3.0     1.0                    23                          18   
        2.0                    13                          15   

                 peripheral pulse  mucous membranes  capillary refill time  \
outcome surgery                                                              
1.0     1.0                    75                78                     84   
        2.0                    64                72                     75   
2.0     1.0                    42                49                     50   
        2.0                    17                17                     17   
3.0     1.0                    17                21                     24   
        2.0                    15                15                     17   

                 pain  ...  abdomen  packed cell volume  total protein  \
outcome surgery        ...                                               
1.0     1.0        75  ...       58                  88             87   
        2.0        69  ...       47                  77             78   
2.0     1.0        48  ...       31                  51             49   
        2.0        17  ...       13                  18             16   
3.0     1.0        20  ...       19                  22             22   
        2.0        15  ...       13                  14             14   

                 abdominocentesis appearance  abdomcentesis total protein  \
outcome surgery                                                             
1.0     1.0                               48                           32   
        2.0                               26                           20   
2.0     1.0                               31                           20   
        2.0                                5                            4   
3.0     1.0                               15                           17   
        2.0                                9                            8   

                 surgical lesion  type of lesion(25)  type of lesion(26)  \
outcome surgery                                                            
1.0     1.0                   95                  95                  95   
        2.0                   83                  83                  83   
2.0     1.0                   58                  58                  58   
        2.0                   19                  19                  19   
3.0     1.0                   27                  27                  27   
        2.0                   17                  17                  17   

                 type of lesion(27)  cp_data  
outcome surgery                               
1.0     1.0                      95       95  
        2.0                      83       83  
2.0     1.0                      58       58  
        2.0                      19       19  
3.0     1.0                      27       27  
        2.0                      17       17  

[6 rows x 26 columns]

In [41]:
lesion = horse_data.groupby('type of lesion(25)').count()
lesion.head(20)

surgery  age  hospital number  rectal temperature  pulse  \
type of lesion(25)                                                             
0                        56   56               56                  51     52   
300                       1    1                1                   1      1   
400                       5    5                5                   4      5   
1111                      1    1                1                   1      1   
1124                      1    1                1                   0      1   
1400                      8    8                8                   6      8   
2111                      3    3                3                   2      3   
2112                      5    5                5                   4      5   
2113                      6    6                6                   6      6   
2124                      9    9                9                   6      8   
2205                     13   13               13                  11     12   
2206                      4    4                4                   4      4   
2207                      3    3                3                   2      2   
2208                     20   20               20                  15     18   
2209                     11   11               11                   9     11   
2300                      1    1                1                   1      1   
2305                      1    1                1                   1      1   
2322                      2    2                2                   2      2   
3025                      2    2                2                   0      0   
3111                     32   33               33                  28     32   

                    respiratory rate  temperature of extremities  \
type of lesion(25)                                                 
0                                 48                          47   
300                                1                           1   
400                                5                           5   
1111                               0                           1   
1124                               1                           1   
1400                               7                           6   
2111                               3                           3   
2112                               4                           4   
2113                               5                           2   
2124                               8                           8   
2205                              13                          13   
2206                               3                           3   
2207                               1                           3   
2208                              16                          13   
2209                               8                          10   
2300                               1                           1   
2305                               1                           1   
2322                               1                           1   
3025                               0                           2   
3111                              24                          27   

                    peripheral pulse  mucous membranes  capillary refill time  \
type of lesion(25)                                                              
0                                 43                48                     49   
300                                1                 1                      1   
400                                5                 5                      5   
1111                               1                 1                      1   
1124                               1                 1                      1   
1400                               6                 7                      8   
2111                               2                 3                      3   
2112                             

In [5]:
req_data = horse_data.loc[:, ['surgery','age','rectal temperature','pulse','abdominal distension', 'abdomcentesis total protein', 'outcome', 'surgical lesion', 'type of lesion(25)']]
req_data.head()

surgery  age  rectal temperature  pulse  abdominal distension  \
0      2.0    1                38.5   66.0                   4.0   
1      1.0    1                39.2   88.0                   2.0   
2      2.0    1                38.3   40.0                   1.0   
3      1.0    9                39.1  164.0                   4.0   
4      2.0    1                37.3  104.0                   NaN   

   abdomcentesis total protein  outcome  surgical lesion  type of lesion(25)  
0                          NaN      2.0                2               11300  
1                          2.0      3.0                2                2208  
2                          NaN      1.0                2                   0  
3                          5.3      2.0                1                2208  
4                          NaN      2.0                2                4300

In [6]:
# рассчитаем базовые метрики (при этом следует учитывать, что ввиду того, что большая часть исходных данных имеет тип категориальных порядковых, базовые метрики не показательны)
req_data.describe()

surgery         age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   

       abdominal distension  abdomcentesis total protein     outcome  \
count            244.000000                   102.000000  299.000000   
mean               2.266393                     3.019608    1.551839   
std                1.065131                     1.968567    0.737187   
min                1.000000                     0.100000    1.000000   
25%                1.000000                     2.000000    1.000000   
50%                2.000000                     2.250000    1.000000   
75%                3.000000                     3.900000    2.000000   
max                4.000000                    10.100000    3.000000   

       surgical lesion  type of lesion(25)  
count       300.000000          300.000000  
mean          1.363333         3657.880000  
std           0.481763         5399.513513  
min           1.000000            0.000000  
25%           1.000000         2111.750000  
50%           1.000000         2673.500000  
75%           2.000000         3209.000000  
max           2.000000        41110.000000

In [15]:
# рассчитаем моду
def moda_calculate ():
    surgery_counts = {}
    for x in req_data.surgery.round():
        if x not in surgery_counts:
            surgery_counts[x] = 1
        else:
            surgery_counts[x] += 1
            
    max_x = 0
    mode_surgery = None
    for k, v in surgery_counts.items():
        if max_x < v:
            max_x = v
            mode_surgery = k
    print('Значение моды для surgery:', mode_surgery, 'количество встречаемости:', max_x)
    
    age_counts = {}
    for x in req_data.age.round():
        if x not in age_counts:
            age_counts[x] = 1
        else:
            age_counts[x] += 1
            
    max_x = 0
    mode_age = None
    for k, v in age_counts.items():
        if max_x < v:
            max_x = v
            mode_age = k
    print ('Значение моды для age:', mode_age, 'количество встречаемости:', max_x)
    
    temperature_counts = {}
    for x in req_data['rectal temperature'].round():
        if x not in temperature_counts:
            temperature_counts[x] = 1
        else:
            temperature_counts[x] += 1
            
    max_x = 0
    mode_temperature = None
    for k, v in temperature_counts.items():
        if max_x < v:
            max_x = v
            mode_temperature = k
    print('Значение моды для rectal temperature:', mode_temperature, 'количество встречаемости:', max_x)
    
    pulse_counts = {}
    for x in req_data.pulse.round():
        if x not in pulse_counts:
            pulse_counts[x] = 1
        else:
            pulse_counts[x] += 1
            
    max_x = 0
    mode_pulse = None
    for k, v in pulse_counts.items():
        if max_x < v:
            max_x = v
            mode_pulse = k
    print('Значение моды для pulse:', mode_pulse, 'количество встречаемости:', max_x)
    
    abdominal_distension_counts = {}
    for x in req_data['abdominal distension'].round():
        if x not in abdominal_distension_counts:
            abdominal_distension_counts[x] = 1
        else:
            abdominal_distension_counts[x] += 1
            
    max_x = 0
    mode_abdominal_distension = None
    for k, v in abdominal_distension_counts.items():
        if max_x < v:
            max_x = v
            mode_abdominal_distension = k
    print('Значение моды для abdominal distension:', mode_abdominal_distension, 'количество встречаемости:', max_x)
    
    abdomcentesis_total_protein_counts = {}
    for x in req_data['abdomcentesis total protein'].round():
        if x not in abdomcentesis_total_protein_counts:
            abdomcentesis_total_protein_counts[x] = 1
        else:
            abdomcentesis_total_protein_counts[x] += 1
    
    max_x = 0
    mode_abdomcentesis_total_protein = None
    for k, v in abdomcentesis_total_protein_counts.items():
        if max_x < v:
            max_x = v
            mode_abdomcentesis_total_protein = k
    print('Значение моды для abdomcentesis total protein:', mode_abdomcentesis_total_protein, 'количество встречаемости:', max_x)
    
    outcome_counts = {}
    for x in req_data.outcome.round():
        if x not in outcome_counts:
            outcome_counts[x] = 1
        else:
            outcome_counts[x] += 1
    
    max_x = 0
    mode_outcome = None
    for k, v in outcome_counts.items():
        if max_x < v:
            max_x = v
            mode_outcome = k
    print('Значение моды для outcome:', mode_outcome, 'количество встречаемости:', max_x)
    
    surgical_lesion_counts = {}
    for x in req_data['surgical lesion'].round():
        if x not in surgical_lesion_counts:
            surgical_lesion_counts[x] = 1
        else:
            surgical_lesion_counts[x] += 1
    
    max_x = 0
    mode_surgical_lesion = None
    for k, v in surgical_lesion_counts.items():
        if max_x < v:
            max_x = v
            mode_surgical_lesion = k
    print('Значение моды для surgical lesion:', mode_surgical_lesion, 'количество встречаемости:', max_x)
    
    type_of_lesion_counts = {}
    for x in req_data['type of lesion(25)'].round():
        if x not in type_of_lesion_counts:
            type_of_lesion_counts[x] = 1
        else:
            type_of_lesion_counts[x] += 1
    
    max_x = 0
    mode_type_of_lesion = None
    for k, v in type_of_lesion_counts.items():
        if max_x < v:
            max_x = v
            mode_type_of_lesion = k
    print('Значение моды для type of lesion(25):', mode_type_of_lesion, 'количество встречаемости:', max_x)
    
moda_calculate()

Значение моды для surgery: 1.0 количество встречаемости: 180
Значение моды для age: 1 количество встречаемости: 276
Значение моды для rectal temperature: 38.0 количество встречаемости: 152
Значение моды для pulse: 48.0 количество встречаемости: 28
Значение моды для abdominal distension: 1.0 количество встречаемости: 76
Значение моды для abdomcentesis total protein: 2.0 количество встречаемости: 34
Значение моды для outcome: 1.0 количество встречаемости: 178
Значение моды для surgical lesion: 1 количество встречаемости: 191
Значение моды для type of lesion(25): 0 количество встречаемости: 56


In [8]:
# рассчитаем дисперсию для rectal temperature, pulse, abdomcentesis total protein

print('rectal temperature')
print(np.var(req_data['rectal temperature'], ddof=1))
print(req_data['rectal temperature'].var())

print('pulse')
print(np.var(req_data.pulse, ddof=1))
print(req_data.pulse.var())

print('abdomcentesis total protein')
print(np.var(req_data['abdomcentesis total protein'], ddof=1))
print(req_data['abdomcentesis total protein'].var())

rectal temperature
0.5362466875871686
0.5362466875871686
pulse
819.7087747035575
819.7087747035575
abdomcentesis total protein
3.8752552902349024
3.8752552902349024


In [9]:
# межквартильный размах

Q1 = req_data['rectal temperature'].quantile(0.25)
Q3 = req_data['rectal temperature'].quantile(0.75)
IQR = Q3 - Q1
print('межквартильный размах для rectal temperature:', IQR)

Q1 = req_data.pulse.quantile(0.25)
Q3 = req_data.pulse.quantile(0.75)
IQR = Q3 - Q1
print('межквартильный размах для pulse', IQR)

Q1 = req_data['abdomcentesis total protein'].quantile(0.25)
Q3 = req_data['abdomcentesis total protein'].quantile(0.75)
IQR = Q3 - Q1
print('межквартильный размах для abdomcentesis total protein', IQR)

межквартильный размах для rectal temperature: 0.7000000000000028
межквартильный размах для pulse 40.0
межквартильный размах для abdomcentesis total protein 1.9


# Краткие выводы:
Представлена информация по 300 случаям, при этом исходя из уникальных номеров (hospital number) в исследовании участвовало 284 лошади. 
Большая часть данных имеет категориальный тип и ранжируется по группам.
Для краткого анализа были выбраны следующие столбцы: категориальные порядковые (surgery, age, abdominal distension, abdomcentesis total protein, outcome, surgical lesion, type of lesion(25)) и количественные непрерывные (rectal temperature, pulse).

Из 300 рассмотренных случаев в 180 случаях имела место операция, в 119 - лечение проходило без операции (потери минимальны, данные представлены по 299 случаям из 300). 

В данных есть описка по возрасту, согласно заданных параметров выделяются 2 категории: 1 - Adult и 2 - Young, но в представленном исследовании встречаются значения: 1 и 9. При анализе исходим из предположения, что значение 9 было ошибочно проставлено вместо значения 2.
92% (276) случаев относились к лошадям категории Adult, по 174 (63%) из них была проведена оперирация, по 101 (36,6%) - лечение без операции. Молодых лошадей (в возрасте <6 месяцев) в исследовании не много - 24 (причина такой динамики не очевидна, возможно основная целевая группа - это взрослые лошади, или маленькие жеребята меньше подвержены болезням, или по данной категории исследовались преимущественно случаи с операциями - 18 из 24 жеребят были прооперированы). 

Данные по температуре (rectal temperature) представлены только по 240 случаям (т.е. потери составляют 20%, оцениваются как существенные). Нормальной температурой считается температура = 37,8 градусов по Цельсию, в большинстве зафиксированных случаев температура была близка к нормальной(слегка повышенная), о чем свидетельствуют близкие значения меры центральной тенденции: среднее арифметическое = 38,17, медиана = 38,20, мода = 38,0 (для 152 случаев из 240). Также отметим, что болезнь скорее сопровождается повышенной температурой, понижение температуры ниже нормальной (37,8) зафиксировано менее чем в 25% случаев. Среднеквадратическое отклонение оценивается как умеренное (0,73 градуса), однако размах между min и max значениями существенный (5,4 градусов) и требует анализа на выбросы.

Данные по пульсу получены по 276 случаям (т.е. потери составляют 8%, предварительно оцениваются как умеренные), нормальным для лошадей в категории Adult признается пульс в пределах 30-40 ударов, в большинстве зафиксированных случаев пульс при болезни существенно повышен, о чем свидетельствуют все рассчитанные показатели. Следует отметить довольно существенный разброс мер центральной тенденции: среднее арифметическое = 72, медианное = 64, мода = 48 (но только для 28 случаев).

Одним из важных параметров при анализе заболеваемости считается показатель вздутия живота (abdominal distention), он был оценен для 244 случаев из 300, в остальных 56 случаях данных нет, потери существенные. Несмотря на его важность, довольно часто данный признак отсутствовал, что подтверждает значение моды = 1 - none в 76 случаях, еще в 65 случаях зафиксировано слабое вздутие (2 - slight) и только в 38 случаях зафиксировано существенное вздутие (4 - severe), свидетельствующее о том, что скорее всего нужна операция.

Показатель уровня протеина (abdomcentesis total protein), позволяющий определить вероятность нарушений кишечника зафиксирован только в 102 случаях, потери крайне высокие, причины требуют доп анализа. В связи с очень большим разбросом в данных (10 единиц между max и min) среднеарифметическое, а также медианное значения представляются нерепрезентативными, в связи с этим наиболее показательным представляется мода, которая равна 2.0 для 34 случаев из 102.

Результат лечения / операции зафиксирован для 299 случаев, 59,5% случаев (178) завершились благополучно (лошадь выжила), в т.ч. для 95 прооперированных лошадей. В 121 случае болезнь завершилась смертью или усыплением, смертность крайне высокая. Из данных 121 случаев операция имела место быть в 85 и, очевидно, не дала положительных результатов. Предварительно можно заключить, что операция помогает только в некоторых случаях (нужна доп оценка всех показателей в комплексе) или на ранних стадиях.

Диагноз (type of lesion(25)) не определен в 56 случаях из 300 (18,7% всех случаев), диагностика оставляет желать лучшего. 



Задание 2. Работа с выбросами
В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [10]:
for col in req_data.columns:
    pct_missing = req_data[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery - 0.3%
age - 0.0%
rectal temperature - 20.0%
pulse - 8.0%
abdominal distension - 18.7%
abdomcentesis total protein - 66.0%
outcome - 0.3%
surgical lesion - 0.0%
type of lesion(25) - 0.0%
